src/test/fuzz/utxo_total_supply.cpp

```diff
@@ -0,0 +1,158 @@
+#include <validation.h>
+#include <version.h>
+
+enum class Action : uint8_t {
+    CREATE_INPUT, //!< Append an input-output pair to the last tx in the current block
+    CREATE_TX,    //!< Append a tx to the list of txs in the current block
+    CREATE_BLOCK, //!< Append the current block to the active chain
+};
+
+
+void test_one_input(const std::vector<uint8_t>& buffer)
+{
+    /** The testing setup that creates a chainstate and other globals */
+    TestingSetup test_setup{CBaseChainParams::REGTEST};
+    FuzzedDataProvider fuzzed_data_provider(buffer.data(), buffer.size());
+
+    const auto PrepareNextBlock = [&]() {
+        // Use OP_FALSE to avoid BIP30 check from hitting early
+        auto block = PrepareBlock(test_setup.m_node, /* coinbase_scriptPubKey */ OP_FALSE);
+        // Replace OP_FALSE with OP_TRUE
+        {
+            CMutableTransaction tx{*block->vtx.back()};
+            tx.vout.at(0).scriptPubKey = OP_TRUE;
+            block->vtx.back() = MakeTransactionRef(tx);
+        }
+        return block;
+    };
+
+    /** The block template this fuzzer is working on */
+    auto current_block = PrepareNextBlock();
+    /** Append-only set of tx outpoints, entries are not removed when spent */
+    std::vector<std::pair<COutPoint, CTxOut>> txos;
+    /** The utxo stats at the chain tip */
+    CCoinsStats utxo_stats;
+    /** The total amount of coins in the utxo set */
+    CAmount circulation{0};
+
+
+    // Store the tx out in the txo map
+    const auto StoreLastTxo = [&]() {
+        // get last tx
+        const CTransaction& tx = *current_block->vtx.back();
+        // get last out
+        const uint32_t i = tx.vout.size() - 1;
+        // store it
+        txos.emplace_back(COutPoint{tx.GetHash(), i}, tx.vout.at(i));
+        if (current_block->vtx.size() == 1 && tx.vout.at(i).scriptPubKey[0] == OP_RETURN) {
+            // also store coinbase
+            const uint32_t i = tx.vout.size() - 2;
+            txos.emplace_back(COutPoint{tx.GetHash(), i}, tx.vout.at(i));
+        }
+    };
+    const auto AppendRandomTxo = [&](CMutableTransaction& tx) {
+        const auto& txo = txos.at(fuzzed_data_provider.ConsumeIntegralInRange<size_t>(0, txos.size() - 1));
+        tx.vin.emplace_back(txo.first);
+        tx.vout.emplace_back(txo.second.nValue, txo.second.scriptPubKey); // "Forward" coin with no fee
+    };
+    const auto UpdateUtxoStats = [&]() {
+        LOCK(cs_main);
+        ::ChainstateActive().ForceFlushStateToDisk();
+        assert(GetUTXOStats(&ChainstateActive().CoinsDB(), utxo_stats));
+        // Check that miner can't print more money than they are allowed to
+        assert(circulation == utxo_stats.nTotalAmount);
+    };
+
+
+    // Update internal state to chain tip
+    StoreLastTxo();
+    UpdateUtxoStats();
+    assert(ChainActive().Height() == 0);
+    // Get at which height we duplicate the coinbase
+    int64_t duplicate_coinbase_height;
+    CScript duplicate_coinbase_script;
+    {
```

Comment:
this section needs a description what it's setting up for

Reply:
Made scope shorter